### PR TITLE
explicit registering of notifiers with watcher

### PR DESCRIPTION
**Pull request by eikenb**
- adds a 'name' field to templates that gets appended to the ID to ensure
  they are unique and can be added the the watche
- 'name' field is optional, ID defaults to just the md5hex if not set
- explicit Register call to the watcher with checks to ensure each Notifier
  is unique and returns an error if registering a duplicate
- attempting to use unregistered Notifier results in a panic
- changes the old Register/register methods to Track/track which more
  closely aligns with their use (and freed up Register)

These semantics should allow multiple parallel uses of the same template
content, with different instances to work while letting the developer
structure the code to ensure everything lines up.

It eliminates the "auto-registering on first use" that was used before
as there was no way for it to properly check (and provide an error) that
the Notifier was a duplicate.

### CODE REVIEW

File: template.go [hunk 1]
```diff
@@ -121,16 +130,31 @@ func NewTemplate(i TemplateInput) *Template {
 	t.dirty = make(drainableChan, 1)
 	t.Notify(nil) // prime template as needing to be run
 
+	t.name = i.Name
+	if t.name == "" {
+		t.name = randStr(randNameLength)
```
Comment (findkim) [inline template.go:135]: The default value for the template name may be difficult to back track which template it is associated with in relation to the template ID. And since the name is autogenerated, it would not be clear which identifier to use for logging purposes: template name or ID. They kinda end up as two miscellaneous values.

What are your thoughts around having the default name fallback to the template ID? 

In this case, the hash of the content and would only allow 1 nameless template of the same content and the second nameless result in the same template from hcat's perspective. It would preserve an option for the same behavior as now, and offer the new feature to have duplicate templates using different names. The developer could choose which design is appropriate for their application

**Review by findkim — COMMENTED** [review]
The name option for templates is a great addition.

I ran into needing this feature when I had tried working on metrics reporting over on Consul Template https://github.com/hashicorp/consul-template/issues/1381 and https://github.com/hashicorp/consul-template/issues/826#issuecomment-627379559. A template could change content in Consul Template when reloading and then the logging and metric history would then be disassociated with it prior to the reload.

File: template.go (hunk 1 above)
Comment (eikenb) [reply template.go:135]: I went back and forth on this myself. I thought the API would be simpler if it used the name-suffix by default to keep the templates/notifiers distinct between instances. As opposed to having it be optional and needing to document why you might want to set it. OTOH, the code would obviously be simpler the other way. 

I assumed CTS would use explicit names, so it wouldn't be an issue. Do you see it being one?

... https://www.youtube.com/watch?v=7QEB38D0pu4 ...

Pondering this more I guess I'm leaning toward your line of thought. Given CTS doesn't need (or want) the random names and all the other current use cases are fine with just the hex-id as the id... I think I'm trying to predict what will be useful the future, which is pretty much always a bad call. 

TLDR; I'm going to strip out the random default, add some documentation and see how things go.

File: watcher.go [hunk 2]
```diff
@@ -249,18 +255,29 @@ func (w *Watcher) Buffer(tmplID string) bool {
 	return w.bufferTemplates.Buffer(tmplID)
 }
 
-// Register is used to add dependencies to be monitored by the watcher. It sets
+// Register's one or more Notifiers with the Watcher for future use.
+// Tring to register the same Notifier twice will result in an error and none
```
Comment (lornasong) [inline watcher.go:257]: minor nit: I think "Tring" might be "Trying"?

File: watcher.go [hunk 3]
```diff
@@ -147,7 +151,9 @@ func (w *Watcher) WatchVaultToken(token string) error {
 			return errors.Wrap(err, "watcher")
 		}
 		// fakeNotifier is defined near end of file
-		w.Register(fakeNotifier(vaultTokenDummyTemplateID), vt)
+		n := fakeNotifier(vaultTokenDummyTemplateID)
+		w.Register(n)
```
Comment (lornasong) [inline watcher.go:155]: Just wanted to make sure I understood here. `w.Register` was updated to return an error. I was wondering if you were considering checking the error here now that it is available? I might be misunderstanding the intention of Register()

Comment (eikenb) [reply watcher.go:155]: Register can return an error, but it requires a setup that makes it easy to not trigger. It is designed more for code where you might be registering templates in different spots, to make sure you don't double register. In tests you have a fresh watcher with only those Register calls you just made right there.

TLDR; The error is checked in the tests that exercise that functionality, but it would just be a lot of line noise in the other tests.

Comment (eikenb) [reply watcher.go:155]: Lol... ignore that. I was just looking at tests and was thinking that was in a test as well. It is not. I'll look at it more closely.

Comment (eikenb) [reply watcher.go:155]: That probably should check the returned error there. I was thinking it not being an issue as you'd only ever need one of those vault token watcher, but you certainly could call that method twice by mistake...

File: watcher.go (hunk 2 above)
Comment (eikenb) [reply watcher.go]: fixed

File: watcher.go (hunk 3 above)
Comment (eikenb) [reply watcher.go]: fixed.

Comment (lornasong) [reply watcher.go]: :) I gotcha!

**Review by lornasong — APPROVED** [review]
makes sense to me! looking forward to using the new template name

File: template.go (hunk 1 above)
Comment (findkim) [reply template.go]: LOL the video caught me off guard! 😂 this was a fun comment to return back to.

Yeah, CTS would use explicit names and the ID+name suffix wouldn't be a problem.

For the Consul Template logging/(future)metric use-case, I anticipate 2 things following the proposal:
1. Logging/metrics based on the name instead of ID would be ideal to keep the same template (updated after reloading changes from disk) would be attributed to the same aggregated bin.
2. Different and unrelated templates can have the same name and it would be up to the caller to unsure it's unique.
